django/contrib/admin/views/main.py

```diff
@@ -35,6 +37,26 @@
     ALL_VAR, ORDER_VAR, ORDER_TYPE_VAR, SEARCH_VAR, IS_POPUP_VAR, TO_FIELD_VAR)
 
 
+class ChangeListForm(forms.Form):
+
+    def __init__(self, *args, **kwargs):
+        super(ChangeListForm, self).__init__(*args, **kwargs)
+        for var in {SEARCH_VAR, PAGE_VAR, TO_FIELD_VAR}:
+            if var == PAGE_VAR:
+                field = forms.IntegerField()
+            else:
+                field = forms.CharField(validators=[ProhibitNullCharactersValidator])
```

Comment:
[This validator is added by default, no need to specify it](https://github.com/django/django/blob/ac956dae1d06ce2ebff7a2966bcaf8a5ecdbb861/django/forms/fields.py#L219). We'll probably want to pass `strip=False` as well to preserve backward compatiblity.

Given you have to branch on `var == PAGE_VAR` anyway I don't think there's much value in using a loop here. You could define fields by directly assigning to `self.fields`

```python
def __init__(self, *args, **kwargs):
    self.fields = {
        SEARCH_VAR: forms.CharField(required=False, strip=False, initial=''),
        PAGE_VAR: forms.IntegerField(required=False, min_value=0, initial=0),
        TO_FIELD_VAR: forms.CharField(required=False),
    }
    super(ChangeListForm, self).__init__()
```

The `initial` usage will make the `clean()` below unnecessary.
```

Reply:
I understood now that ProhibitNullCharactersValidator was appended as CharField's default validator. I missed it. 
your code is very clear and wisely. I applied it